### PR TITLE
Add reference to a utility script to compare keychain dump

### LINKED ISSUE #1790: Keychain Data Persistence: Utility script proposal
Hello,

In order to perform the verification about the **[Keychain Data Persistence](https://github.com/OWASP/owasp-mstg/blob/master/Document/0x06d-Testing-Data-Storage.md#keychain-data-persistence)**, I have created the following script to automate the comparison after the export of the keychain using Objection (like mentioned in the MSTG).

I do not know if you can find it useful for this part of the MSTG so I open a issue just to propose it to you:

**Script:**

```python
import json
import binascii
import hashlib
import argparse
from tabulate import tabulate

"""
Python3 script to find common entries in 2 export of a iOS device keychain performed via objection.

The objective is to help performing the following test of the OWASP MSTG:
https://github.com/OWASP/owasp-mstg/blob/master/Document/0x06d-Testing-Data-Storage.md#keychain-data-persistence

Dependencies:
    pip install tabulate
"""

def search_entry(data_hex, entries):
    """
    Search for a keychain entry in the JSON export of the keychain obtained after 
    the reinstall of the application. Search is performed based on HEX data.
    """
    entry_found_data = None
    raw_data = binascii.unhexlify(data_hex.strip())
    hash_data = hashlib.sha256(raw_data).hexdigest()
    for entry in entries:
        entry_hex_data = entry["dataHex"].strip()
        if len(entry_hex_data) > 0:
            entry_raw_data =  binascii.unhexlify(entry_hex_data)
            entry_hash_data = hashlib.sha256(entry_raw_data).hexdigest()
            if entry_hash_data == hash_data:
                entry_found_data = entry
                break
    return entry_found_data 

parser = argparse.ArgumentParser(description="Script to find common entries in 2 export of a iOS device keychain performed via objection.")
parser.add_argument("--keychain-before", action="store", dest="keychain_export_file_initial_install", help="JSON export of the keychain from the intial installation of the application.", required=True)
parser.add_argument("--keychain-after", action="store", dest="keychain_export_file_after_reinstall", help="JSON export of the keychain after the reinstallation of the application.", required=True)
args = parser.parse_args()

print("[+] Load keychains files...")
with open(args.keychain_export_file_initial_install) as json_file:
    keychain_export_json_initial_install = json.load(json_file)
with open(args.keychain_export_file_after_reinstall) as json_file:
    keychain_export_json_after_reinstall = json.load(json_file)
print(f"{args.keychain_export_file_initial_install}: {len(keychain_export_json_initial_install)} entries")
print(f"{args.keychain_export_file_after_reinstall}: {len(keychain_export_json_after_reinstall)} entries")

print("[+] Search for keychain entries from initial install still present after app reinstall...")
common_entries_found = []
for entry in keychain_export_json_initial_install:
    entry_in_new_keychain = search_entry(entry["dataHex"], keychain_export_json_after_reinstall)
    if entry_in_new_keychain is not None:
        common_entries_found.append(entry_in_new_keychain)

if len(common_entries_found) == 0:
    print("[!] No common entries found.")
else:        
    print("[+] Common entries found:")
    table_headers = ["Account", "DataHex[:20]", "CreateDate", "ModificationDate", "AccessibleAttribute", "ItemClass", "EntitlementGroup"]
    table_rows = []
    for entry in common_entries_found:
        table_rows.append([entry["account"], entry["dataHex"][:20], entry["create_date"].replace("+0000",""), entry["modification_date"].replace("+0000",""), entry["accessible_attribute"], 
        entry["item_class"], entry["entitlement_group"] ])
    print(tabulate(table_rows, headers=table_headers))
```

**Call:**

```shell
$ python keychain-diff.py --help
usage: keychain-diff.py [-h] --keychain-before KEYCHAIN_EXPORT_FILE_INITIAL_INSTALL  --keychain-after KEYCHAIN_EXPORT_FILE_AFTER_REINSTALL

Script to find common entries in 2 export of a iOS device keychain performed
via objection.

optional arguments:
  -h, --help            show this help message and exit
  --keychain-before KEYCHAIN_EXPORT_FILE_INITIAL_INSTALL
                        JSON export of the keychain from the intial
                        installation of the application.
  --keychain-after KEYCHAIN_EXPORT_FILE_AFTER_REINSTALL
                        JSON export of the keychain after the reinstallation
                        of the application.
``` 

Feel free to close the issue if you find it not relevant.

Thank you very much in advance 😃

### PR DESCRIPTION
**Pull request by righettod**
Hi,

Following our discution, this is the PR for the issue #1790 

- [x] Your contribution is written in the 2nd person (e.g. you)
- [x] Your contribution is written in an active present form for as much as possible.
- [x] You have made sure that the reference section is up to date (e.g. please add sources you have used, make sure that the references to MITRE/MASVS/etc. are up to date)
- [x] Your contribution has proper formatted markdown and/or code
- [x] Any references to website have been formatted as [TEXT](URL “NAME”)
- [x] You verified/tested the effectiveness of your contribution (e.g.: is the code really an effective remediation? Please verify it works!)

This PR closes #1790

### CODE REVIEW

**Comment by righettod** [comment]
The PR fail but i'm not the root case 😃 

image

**Comment by commjoen** [comment]
@sushi2k / @cpholguera : i guess we can remove that right?

**Comment by cpholguera** [comment]
Hi @commjoen , do you mean that link? If we're not using that survey we should remove the sentence referencing it. If not we should create a new one and update the URL, it does in fact not exist anymore.

**Comment by commjoen** [comment]
I don't think we are using it, feel free to remove it completely (Sentence+link)

File: Document/0x06d-Testing-Data-Storage.md [hunk 1]
```diff
@@ -103,7 +103,37 @@ Dumping the iOS keychain...
 Created                    Accessible                      ACL    Type      Account                    Service                                                        Data
 -------------------------  ------------------------------  -----  --------  -------------------------  -------------------------------------------------------------  ------------------------------------
 2020-02-11 13:26:52 +0000  WhenUnlocked                    None   Password  keychainValue              com.highaltitudehacks.DVIAswiftv2.develop                      mysecretpass123
-```  
+```
+
+The dedicated script referenced in the [tools](#tools) section can be used to find common entries between two Keychain dump.
```
Comment (TheDauntless) [inline Document/0x06d-Testing-Data-Storage.md:108]:
```suggestion
The dedicated script referenced in the [tools](#tools) section can be used to find common entries between two Keychain dumps.
```

File: Document/0x06d-Testing-Data-Storage.md [hunk 2]
```diff
@@ -103,7 +103,37 @@ Dumping the iOS keychain...
...
+
+Use it like this:
+
+1. Install the application and use it (inital install).
+2. Dump the Keychain using `objection` to a json file.
+3. Remove and reinstall the application.
+4. Dump the Keychain using `objection` to a json file.
```
Comment (TheDauntless) [inline Document/0x06d-Testing-Data-Storage.md:115]:
```suggestion
4. Dump the Keychain using `objection` to `DumpKeyChainAfterReinstall.json`
```

File: Document/0x06d-Testing-Data-Storage.md [hunk 3]
```diff
@@ -103,7 +103,37 @@ Dumping the iOS keychain...
...
 2020-02-11 13:26:52 +0000  WhenUnlocked                    None   Password  keychainValue              com.highaltitudehacks.DVIAswiftv2.develop                      mysecretpass123
-```  
+```
+
+The dedicated script referenced in the [tools](#tools) section can be used to find common entries between two Keychain dump.
+
+Use it like this:
+
+1. Install the application and use it (inital install).
+2. Dump the Keychain using `objection` to a json file.
```
Comment (TheDauntless) [inline Document/0x06d-Testing-Data-Storage.md:113]:
```suggestion
2. Dump the Keychain using `objection` to `DumpKeyChainAfterInitialInstall,json`
```

File: Document/0x06d-Testing-Data-Storage.md [hunk 4]
```diff
@@ -103,7 +103,37 @@ Dumping the iOS keychain...
...
 -------------------------  ------------------------------  -----  --------  -------------------------  -------------------------------------------------------------  ------------------------------------
 2020-02-11 13:26:52 +0000  WhenUnlocked                    None   Password  keychainValue              com.highaltitudehacks.DVIAswiftv2.develop                      mysecretpass123
-```  
+```
+
+The dedicated script referenced in the [tools](#tools) section can be used to find common entries between two Keychain dump.
+
+Use it like this:
+
+1. Install the application and use it (inital install).
```
Comment (TheDauntless) [inline Document/0x06d-Testing-Data-Storage.md:112]:
```suggestion
1. Install the application and use it (initial install).
```

File: Document/0x06d-Testing-Data-Storage.md (hunk 3 above)
Comment (righettod) [reply Document/0x06d-Testing-Data-Storage.md]: I have also fixed the typo `DumpKeyChainAfterInitialInstall,json` by `DumpKeyChainAfterInitialInstall.json`

**Comment by righettod** [comment]
@TheDauntless I have accepted all the proposed update and fixed a small typo.

**Comment by righettod** [comment]
Any news about the merging or rejection?

**Comment by cpholguera** [comment]
Hi @righettod, sorry that it took some time, but we could finally discuss this in our last meeting and tend to think that doing a diff (+eventually using other bash _foo_ e.g. grep) is much more flexible, you can display not only what the dumps have in common but also where they differ. And also important, we don't create a new tool dependency.

We hope you understand this and would still like to thank you a lot for taking the time to write everything 😊.

**Comment by cpholguera** [comment]
Closing after https://github.com/OWASP/owasp-mstg/pull/1792#issuecomment-670841647. Thanks @righettod and if you have any questions feel free to contact us any time.

**Comment by righettod** [comment]
Thanks for the feedback and sorry for the delay of the reply, as I have mentioned in the initial issue, it was just a proposal 😃 
So I perfectly understand.

**Pull request closed without merging**